aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+    elif isinstance(t, OptionalType):
+        # TODO: these arguments are smoothed over by the hacky wrapper
+        if argument and legacy_optional and str(t.elem) == 'Tensor':
+            if mutable:
+                return 'Tensor &'
+            else:
+                return 'const Tensor &'
+        if argument and str(t.elem) == 'Tensor':
+            if mutable:
+                return 'Tensor &'  # WUUUT
+            else:
+                return 'const c10::optional<Tensor>&'
+        elem = cpp_type(t.elem, mutable=mutable, argument=argument, legacy_optional=legacy_optional)
+        return f"c10::optional<{elem}>"
+    elif isinstance(t, ListType):
+        if str(t.elem) == 'bool':
+            assert t.size is not None
+            return f"std::array<bool,{t.size}>"
+        # TODO: remove this special case
+        if str(t.elem) == 'int' and argument:
+            return "IntArrayRef"
+        elif str(t.elem) == 'Tensor' and argument:
+            return "TensorList"
+        elif str(t.elem) == 'Tensor?' and argument and legacy_optional:
+            return "TensorList"
+        elif str(t.elem) == 'Dimname' and argument:
+            return "DimnameList"
+        elem = cpp_type(t.elem, mutable=mutable, argument=argument, legacy_optional=legacy_optional)
+        if argument:
+            # TODO: explicitly qualify namespace here
+            return f"ArrayRef<{elem}>"
+        else:
+            assert t.size is None, f"fixed size list returns not supported: {t}"
+            return f"std::vector<{elem}>"
+    else:
+        assert False
+
+def cpp_type_return(rs: Sequence[Argument]) -> str:
+    if len(rs) == 0:
+        return 'void'
+    elif len(rs) == 1:
+        return cpp_type(rs[0].type, mutable=rs[0].is_write, argument=False, legacy_optional=False)
+    else:
+        args = ','.join([cpp_type(r.type, mutable=r.is_write, argument=False, legacy_optional=False) for r in rs])
+        return f'std::tuple<{args}>'
+
+# Some simple code to exercise some of the functions we've been building
+def compute_type_derived_method_declarations(dispatch: str) -> List[str]:
+    type_derived_method_declarations: List[str] = []
+    for f in native_functions:
+        if f.dispatch is None or dispatch not in f.dispatch:
+            continue
+
+        name = str(f.func.name.name)
+        # TODO: delete this!
+        if f.func.is_out_fn():
+            name += '_out'
+        if f.func.name.overload_name:
+            name += f'_{f.func.name.overload_name}'
+
+        cpp_return = cpp_type_return(f.func.returns)
+
+        def format_arg(a: Argument) -> str:
+            return f"{cpp_type(a.type, mutable=a.is_write, argument=True, legacy_optional=True)} {a.name}"
+        cpp_args: List[str] = []
+        cpp_args.extend(map(format_arg, f.func.out_arguments))
+        cpp_args.extend(map(format_arg, f.func.arguments))
+
+        # Discover TensorOptions
+        topt_names = ['dtype', 'layout', 'device', 'pin_memory']
```

Comment:
Also personally I tried in `gen_python_functions` to always hoist these out to globals and capitalize the names - not so much for efficiency as to visually identify the constants that were driving the proceedings